agent/commands/run.go

```diff
@@ -56,75 +55,89 @@ func Run() {
 		s := <-signals
 		signal.Stop(signals)
 		l.Warnf("Got %s, shutting down...", unix.SignalName(s.(unix.Signal))) //nolint:forcetypeassert
-		if cfg != nil {
-			cleanupTmp(cfg.Paths.TempDir, l)
-		}
-		cancel()
+		rootCancel()
 	}()
 
-	configStorage := config.NewStorage(nil)
-	configFilepath, err := configStorage.Reload(l)
-	if err != nil {
-		l.Fatalf("Failed to load configuration: %s.", err)
-	}
-
-	cfg = configStorage.Get()
-
-	cleanupTmp(cfg.Paths.TempDir, l)
-	connectionUptimeService := connectionuptime.NewService(cfg.WindowConnectedTime)
-	connectionUptimeService.RunCleanupGoroutine(ctx)
 	v := versioner.New(&versioner.RealExecFunctions{})
-	supervisor := supervisor.NewSupervisor(ctx, v, configStorage)
-	connectionChecker := connectionchecker.New(configStorage)
-	r := runner.New(cfg.RunnerCapacity)
-	client := client.New(configStorage, supervisor, r, connectionChecker, v, connectionUptimeService, logStore)
-	localServer := agentlocal.NewServer(configStorage, supervisor, client, configFilepath, logStore)
-
-	var wg sync.WaitGroup
-	wg.Add(3)
-	go func() {
-		defer wg.Done()
-		supervisor.Run(ctx)
-		cancel()
-	}()
-	go func() {
-		defer wg.Done()
-		r.Run(ctx)
-		cancel()
-	}()
-	go func() {
-		defer wg.Done()
-		localServer.Run(ctx)
-		cancel()
-	}()
+	configStorage, configFilepath := prepareConfig(l)
 
 	for {
-		_, err = configStorage.Reload(l)
-		if err != nil {
-			l.Fatalf("Failed to load configuration: %s.", err)
-		}
-
+		ctx, cancel := context.WithCancel(rootCtx)
 		cfg := configStorage.Get()
 
-		config.ConfigureLogger(cfg)
-		logStore.Resize(cfg.LogLinesCount)
-		l.Debugf("Loaded configuration: %+v", cfg)
+		prepareLogger(cfg, logStore, l)
+
+		supervisor := supervisor.NewSupervisor(ctx, v, configStorage)
+		connectionChecker := connectionchecker.New(configStorage)
+		r := runner.New(cfg.RunnerCapacity)
+		client := client.New(configStorage, supervisor, r, connectionChecker, v, prepareConnectionService(ctx, cfg), logStore)
+		localServer := agentlocal.NewServer(configStorage, supervisor, client, configFilepath, logStore)
 
 		logrus.Infof("Window check connection time is %.2f hour(s)", cfg.WindowConnectedTime.Hours())
-		connectionUptimeService.SetWindowPeriod(cfg.WindowConnectedTime)
 
+		var wg sync.WaitGroup
+		wg.Add(3)
+		reloadCh := make(chan bool, 1)
+		go func() {
+			defer wg.Done()
+			supervisor.Run(ctx)
+			cancel()
+		}()
+		go func() {
+			defer wg.Done()
+			r.Run(ctx)
+			cancel()
+		}()
+		go func() {
+			defer wg.Done()
+			localServer.Run(ctx, reloadCh)
+			cancel()
+		}()
+
+		isReload := processClientUntilCancel(ctx, client, reloadCh, l)
+
+		cleanupTmp(cfg.Paths.TempDir, l)
+		wg.Wait()
+		if !isReload {
+			return
+		}
```

Comment:
```suggestion
		select {
		case <- ctx.Done():
		        return
		default:
		}
```

Reply:
With current suggested changes pmm-agent will quit in case of reload. Context is done in case of SIGINT but also in case of reload by config subcommand.

Your suggestion:
		select {
		case <- ctx.Done():
		        return
		default:
		}

With both suggestion above we cannot determine which case it is and it will quit main loop in both cases.

Reply:
As discussed with rootCtx its OK.